File: backend/app/parsers/svg_parser.py
"""
SVG Path Parser (Server-side)
Extracts <path> data from SVG content and samples into discrete (x,y) points.
Handles M, L, H, V, C, Q, A, Z commands (absolute and relative).
"""

import re
import math
from typing import List, Dict, Tuple, Optional
from lxml import etree


def parse_svg(content: str, num_samples: int = 101) -> Dict:
    """
    Parse SVG content and return sampled path points.

    Returns:
        dict with 'points' (list of [x,y]), 'bounds', 'path_d'
    """
    # Parse XML
    try:
        root = etree.fromstring(content.encode("utf-8"))
    except etree.XMLSyntaxError as e:
        raise ValueError(f"Invalid SVG XML: {e}")

    # Handle namespaces
    ns = {"svg": "http://www.w3.org/2000/svg"}
    path_el = root.find(".//svg:path", ns) or root.find(".//path")

    if path_el is None:
        # Try without namespace
        for el in root.iter():
            if el.tag.endswith("path") or el.tag == "path":
                path_el = el
                break

    if path_el is None:
        raise ValueError("No <path> element found in SVG")

    d = path_el.get("d", "")
    if not d.strip():
        raise ValueError("Path d attribute is empty")

    # Parse path commands into segments
    segments = _parse_path_commands(d)

    # Sample into dense points
    dense = _densely_sample(segments)
    if len(dense) < 2:
        raise ValueError(f"Path produced only {len(dense)} points — too few to trace")

    # Resample at uniform arc length
    points = _resample_arc_length(dense, num_samples)

    # Compute bounds
    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    bounds = {
        "minX": min(xs), "minY": min(ys),
        "maxX": max(xs), "maxY": max(ys),
    }

    return {
        "points": points,
        "bounds": bounds,
        "path_d": d,
    }


def _parse_path_commands(d: str) -> list:
    """Tokenize SVG path d-attribute into command segments."""
    segments = []
    tokens = re.findall(r'[a-zA-Z][^a-zA-Z]*', d)
    cx, cy = 0.0, 0.0
    sx, sy = 0.0, 0.0  # subpath start

    for token in tokens:
        cmd = token[0]
        nums = [float(x) for x in re.findall(r'-?[\d.]+(?:e[+-]?\d+)?', token[1:], re.IGNORECASE)]
        is_rel = cmd.islower()
        CMD = cmd.upper()

        if CMD == 'M':
            x = (cx + nums[0]) if is_rel else nums[0]
            y = (cy + nums[1]) if is_rel else nums[1]
            cx, cy = x, y
            sx, sy = x, y
            segments.append(('M', x, y))
            # Implicit lineto for remaining pairs
            for i in range(2, len(nums), 2):
                lx = (cx + nums[i]) if is_rel else nums[i]
                ly = (cy + nums[i+1]) if is_rel else nums[i+1]
                segments.append(('L', cx, cy, lx, ly))
                cx, cy = lx, ly

        elif CMD == 'L':
            for i in range(0, len(nums), 2):
                x = (cx + nums[i]) if is_rel else nums[i]
                y = (cy + nums[i+1]) if is_rel else nums[i+1]
                segments.append(('L', cx, cy, x, y))
                cx, cy = x, y

        elif CMD == 'H':
            for n in nums:
                x = (cx + n) if is_rel else n
                segments.append(('L', cx, cy, x, cy))
                cx = x

        elif CMD == 'V':
            for n in nums:
                y = (cy + n) if is_rel else n
                segments.append(('L', cx, cy, cx, y))
                cy = y

        elif CMD == 'C':
            for i in range(0, len(nums), 6):
                cp1x = (cx + nums[i]) if is_rel else nums[i]
                cp1y = (cy + nums[i+1]) if is_rel else nums[i+1]
                cp2x = (cx + nums[i+2]) if is_rel else nums[i+2]
                cp2y = (cy + nums[i+3]) if is_rel else nums[i+3]
                ex = (cx + nums[i+4]) if is_rel else nums[i+4]
                ey = (cy + nums[i+5]) if is_rel else nums[i+5]
                segments.append(('C', cx, cy, cp1x, cp1y, cp2x, cp2y, ex, ey))
                cx, cy = ex, ey

        elif CMD == 'Q':
            for i in range(0, len(nums), 4):
                cpx = (cx + nums[i]) if is_rel else nums[i]
                cpy = (cy + nums[i+1]) if is_rel else nums[i+1]
                ex = (cx + nums[i+2]) if is_rel else nums[i+2]
                ey = (cy + nums[i+3]) if is_rel else nums[i+3]
                segments.append(('Q', cx, cy, cpx, cpy, ex, ey))
                cx, cy = ex, ey

        elif CMD == 'A':
            for i in range(0, len(nums), 7):
                rx, ry = abs(nums[i]), abs(nums[i+1])
                rot = nums[i+2]
                large = int(nums[i+3])
                sweep = int(nums[i+4])
                ex = (cx + nums[i+5]) if is_rel else nums[i+5]
                ey = (cy + nums[i+6]) if is_rel else nums[i+6]
                segments.append(('A', cx, cy, rx, ry, rot, large, sweep, ex, ey))
                cx, cy = ex, ey

        elif CMD == 'Z':
            if abs(cx - sx) > 1e-6 or abs(cy - sy) > 1e-6:
                segments.append(('L', cx, cy, sx, sy))
            cx, cy = sx, sy

    return segments


def _densely_sample(segments: list, pts_per_curve: int = 50) -> List[List[float]]:
    """Sample each segment densely."""
    points = []

    for seg in segments:
        cmd = seg[0]
        if cmd == 'M':
            points.append([seg[1], seg[2]])

        elif cmd == 'L':
            points.append([seg[3], seg[4]])

        elif cmd == 'C':
            _, x0, y0, cp1x, cp1y, cp2x, cp2y, x1, y1 = seg
            for i in range(1, pts_per_curve + 1):
                t = i / pts_per_curve
                mt = 1 - t
                x = mt**3*x0 + 3*mt**2*t*cp1x + 3*mt*t**2*cp2x + t**3*x1
                y = mt**3*y0 + 3*mt**2*t*cp1y + 3*mt*t**2*cp2y + t**3*y1
                points.append([x, y])

        elif cmd == 'Q':
            _, x0, y0, cpx, cpy, x1, y1 = seg
            for i in range(1, pts_per_curve + 1):
                t = i / pts_per_curve
                mt = 1 - t
                x = mt**2*x0 + 2*mt*t*cpx + t**2*x1
                y = mt**2*y0 + 2*mt*t*cpy + t**2*y1
                points.append([x, y])

        elif cmd == 'A':
            _, x0, y0, rx, ry, rot, large, sweep, x1, y1 = seg
            arc_pts = _approximate_arc(x0, y0, rx, ry, rot, large, sweep, x1, y1, pts_per_curve)
            points.extend(arc_pts[1:])  # Skip first (already in list)

    return points


def _approximate_arc(x1, y1, rx, ry, rotation, large_arc, sweep, x2, y2, n):
    """Approximate SVG arc as polyline using endpoint-to-center conversion."""
    if rx == 0 or ry == 0:
        return [[x1, y1], [x2, y2]]

    phi = math.radians(rotation)
    cos_phi, sin_phi = math.cos(phi), math.sin(phi)

    dx2, dy2 = (x1 - x2) / 2, (y1 - y2) / 2
    x1p = cos_phi * dx2 + sin_phi * dy2
    y1p = -sin_phi * dx2 + cos_phi * dy2

    lam = (x1p**2) / (rx**2) + (y1p**2) / (ry**2)
    if lam > 1:
        s = math.sqrt(lam)
        rx *= s
        ry *= s

    num = max(0, rx**2 * ry**2 - rx**2 * y1p**2 - ry**2 * x1p**2)
    den = rx**2 * y1p**2 + ry**2 * x1p**2
    sq = math.sqrt(num / den) if den > 0 else 0
    if large_arc == sweep:
        sq = -sq

    cxp = sq * (rx * y1p) / ry
    cyp = sq * -(ry * x1p) / rx

    cx = cos_phi * cxp - sin_phi * cyp + (x1 + x2) / 2
    cy = sin_phi * cxp + cos_phi * cyp + (y1 + y2) / 2

    theta1 = math.atan2((y1p - cyp) / ry, (x1p - cxp) / rx)
    dtheta = math.atan2((-y1p - cyp) / ry, (-x1p - cxp) / rx) - theta1

    if sweep and dtheta < 0:
        dtheta += 2 * math.pi
    if not sweep and dtheta > 0:
        dtheta -= 2 * math.pi

    pts = []
    for i in range(n + 1):
        t = theta1 + (i / n) * dtheta
        xr, yr = rx * math.cos(t), ry * math.sin(t)
        pts.append([
            cos_phi * xr - sin_phi * yr + cx,
            sin_phi * xr + cos_phi * yr + cy,
        ])
    return pts


def _resample_arc_length(points: List[List[float]], n: int) -> List[List[float]]:
    """Resample polyline to n uniformly spaced points by arc length."""
    if len(points) < 2:
        return points

    # Cumulative arc lengths
    arc = [0.0]
    for i in range(1, len(points)):
        dx = points[i][0] - points[i-1][0]
        dy = points[i][1] - points[i-1][1]
        arc.append(arc[-1] + math.sqrt(dx*dx + dy*dy))

    total = arc[-1]
    if total == 0:
        return [points[0]]

    result = []
    j = 0
    for i in range(n):
        target = (i / (n - 1)) * total
        while j < len(arc) - 2 and arc[j+1] < target:
            j += 1
        seg_len = arc[j+1] - arc[j]
        frac = (target - arc[j]) / seg_len if seg_len > 0 else 0
        result.append([
            points[j][0] + (points[j+1][0] - points[j][0]) * frac,
            points[j][1] + (points[j+1][1] - points[j][1]) * frac,
        ])

    return result

File: frontend/src/utils/svgParser.js
/**
 * SVG Path Parser
 * Extracts path data from SVG content and samples it into discrete (x,y) points.
 * Handles: M, L, C, Q, A, Z commands (absolute and relative).
 */

/**
 * Parse SVG string content and extract sampled points from <path> elements.
 *
 * @param {string} svgContent - Raw SVG file content
 * @param {number} [numSamples=101] - Number of points to sample along the path
 * @returns {{ points: Array<{x: number, y: number}>, viewBox: {x:number,y:number,w:number,h:number}|null, rawPath: string }}
 */
export function parseSvg(svgContent, numSamples = 101) {
  const parser = new DOMParser();
  const doc = parser.parseFromString(svgContent, 'image/svg+xml');

  // Check for parse errors
  const errorNode = doc.querySelector('parsererror');
  if (errorNode) {
    throw new Error('Invalid SVG: XML parse error');
  }

  const svgEl = doc.querySelector('svg');
  if (!svgEl) throw new Error('No <svg> element found');

  // Extract viewBox
  let viewBox = null;
  const vb = svgEl.getAttribute('viewBox');
  if (vb) {
    const parts = vb.split(/[\s,]+/).map(Number);
    if (parts.length === 4) {
      viewBox = { x: parts[0], y: parts[1], w: parts[2], h: parts[3] };
    }
  }

  // Find the first <path> element (search nested groups too)
  const pathEl = doc.querySelector('path');
  if (!pathEl) throw new Error('No <path> element found in SVG');

  const rawPath = pathEl.getAttribute('d') || '';
  if (!rawPath.trim()) throw new Error('Path d attribute is empty');

  // Parse and sample the path
  const segments = parsePathCommands(rawPath);
  const points = sampleSegments(segments, numSamples);

  return { points, viewBox, rawPath };
}

/**
 * Parse SVG path d-attribute into segment objects.
 */
function parsePathCommands(d) {
  const segments = [];
  // Tokenize: split into commands and their numeric arguments
  const tokens = d.match(/[a-zA-Z][^a-zA-Z]*/g) || [];

  let cx = 0, cy = 0; // Current point
  let sx = 0, sy = 0; // Subpath start

  for (const token of tokens) {
    const cmd = token[0];
    const nums = (token.slice(1).match(/-?[\d.]+(?:e[+-]?\d+)?/gi) || []).map(Number);
    const isRel = cmd === cmd.toLowerCase();
    const CMD = cmd.toUpperCase();

    if (CMD === 'M') {
      // MoveTo
      const x = isRel ? cx + nums[0] : nums[0];
      const y = isRel ? cy + nums[1] : nums[1];
      cx = x; cy = y;
      sx = x; sy = y;
      segments.push({ type: 'M', x, y });
      // Implicit LineTo for remaining pairs
      for (let i = 2; i < nums.length; i += 2) {
        const lx = isRel ? cx + nums[i] : nums[i];
        const ly = isRel ? cy + nums[i + 1] : nums[i + 1];
        segments.push({ type: 'L', x1: cx, y1: cy, x2: lx, y2: ly });
        cx = lx; cy = ly;
      }
    } else if (CMD === 'L') {
      for (let i = 0; i < nums.length; i += 2) {
        const x = isRel ? cx + nums[i] : nums[i];
        const y = isRel ? cy + nums[i + 1] : nums[i + 1];
        segments.push({ type: 'L', x1: cx, y1: cy, x2: x, y2: y });
        cx = x; cy = y;
      }
    } else if (CMD === 'H') {
      for (const n of nums) {
        const x = isRel ? cx + n : n;
        segments.push({ type: 'L', x1: cx, y1: cy, x2: x, y2: cy });
        cx = x;
      }
    } else if (CMD === 'V') {
      for (const n of nums) {
        const y = isRel ? cy + n : n;
        segments.push({ type: 'L', x1: cx, y1: cy, x2: cx, y2: y });
        cy = y;
      }
    } else if (CMD === 'C') {
      // Cubic Bézier
      for (let i = 0; i < nums.length; i += 6) {
        const cp1x = isRel ? cx + nums[i] : nums[i];
        const cp1y = isRel ? cy + nums[i + 1] : nums[i + 1];
        const cp2x = isRel ? cx + nums[i + 2] : nums[i + 2];
        const cp2y = isRel ? cy + nums[i + 3] : nums[i + 3];
        const ex = isRel ? cx + nums[i + 4] : nums[i + 4];
        const ey = isRel ? cy + nums[i + 5] : nums[i + 5];
        segments.push({
          type: 'C', x1: cx, y1: cy,
          cp1x, cp1y, cp2x, cp2y,
          x2: ex, y2: ey,
        });
        cx = ex; cy = ey;
      }
    } else if (CMD === 'Q') {
      // Quadratic Bézier
      for (let i = 0; i < nums.length; i += 4) {
        const cpx = isRel ? cx + nums[i] : nums[i];
        const cpy = isRel ? cy + nums[i + 1] : nums[i + 1];
        const ex = isRel ? cx + nums[i + 2] : nums[i + 2];
        const ey = isRel ? cy + nums[i + 3] : nums[i + 3];
        segments.push({
          type: 'Q', x1: cx, y1: cy,
          cpx, cpy,
          x2: ex, y2: ey,
        });
        cx = ex; cy = ey;
      }
    } else if (CMD === 'A') {
      // Arc
      for (let i = 0; i < nums.length; i += 7) {
        const rx = nums[i], ry = nums[i + 1];
        const rotation = nums[i + 2];
        const largeArc = nums[i + 3];
        const sweep = nums[i + 4];
        const ex = isRel ? cx + nums[i + 5] : nums[i + 5];
        const ey = isRel ? cy + nums[i + 6] : nums[i + 6];
        segments.push({
          type: 'A', x1: cx, y1: cy,
          rx, ry, rotation, largeArc, sweep,
          x2: ex, y2: ey,
        });
        cx = ex; cy = ey;
      }
    } else if (CMD === 'Z') {
      if (Math.abs(cx - sx) > 0.001 || Math.abs(cy - sy) > 0.001) {
        segments.push({ type: 'L', x1: cx, y1: cy, x2: sx, y2: sy });
      }
      cx = sx; cy = sy;
    }
  }

  return segments;
}

/**
 * Sample all path segments into evenly-spaced points.
 */
function sampleSegments(segments, numSamples) {
  // First, densely sample all segments
  const dense = [];
  for (const seg of segments) {
    if (seg.type === 'M') {
      dense.push({ x: seg.x, y: seg.y });
    } else if (seg.type === 'L') {
      // Add endpoint (start already in dense from previous segment)
      dense.push({ x: seg.x2, y: seg.y2 });
    } else if (seg.type === 'C') {
      // Cubic Bézier — sample at 50 points per segment
      for (let i = 1; i <= 50; i++) {
        const t = i / 50;
        dense.push(cubicBezierAt(seg.x1, seg.y1, seg.cp1x, seg.cp1y, seg.cp2x, seg.cp2y, seg.x2, seg.y2, t));
      }
    } else if (seg.type === 'Q') {
      for (let i = 1; i <= 50; i++) {
        const t = i / 50;
        dense.push(quadBezierAt(seg.x1, seg.y1, seg.cpx, seg.cpy, seg.x2, seg.y2, t));
      }
    } else if (seg.type === 'A') {
      // Approximate arc with many line segments
      const arcPts = approximateArc(seg, 50);
      for (let i = 1; i < arcPts.length; i++) {
        dense.push(arcPts[i]);
      }
    }
  }

  if (dense.length < 2) return dense;

  // Compute cumulative arc lengths
  const arcLens = [0];
  for (let i = 1; i < dense.length; i++) {
    const dx = dense[i].x - dense[i - 1].x;
    const dy = dense[i].y - dense[i - 1].y;
    arcLens.push(arcLens[i - 1] + Math.sqrt(dx * dx + dy * dy));
  }
  const totalLen = arcLens[arcLens.length - 1];
  if (totalLen === 0) return [dense[0]];

  // Resample at uniform arc-length intervals
  const result = [];
  for (let i = 0; i < numSamples; i++) {
    const targetLen = (i / (numSamples - 1)) * totalLen;
    // Binary search for segment
    let lo = 0, hi = arcLens.length - 1;
    while (lo < hi - 1) {
      const mid = (lo + hi) >> 1;
      if (arcLens[mid] <= targetLen) lo = mid;
      else hi = mid;
    }
    const segFrac = (arcLens[hi] - arcLens[lo]) > 0
      ? (targetLen - arcLens[lo]) / (arcLens[hi] - arcLens[lo])
      : 0;
    result.push({
      x: dense[lo].x + (dense[hi].x - dense[lo].x) * segFrac,
      y: dense[lo].y + (dense[hi].y - dense[lo].y) * segFrac,
    });
  }

  return result;
}

function cubicBezierAt(x1, y1, cp1x, cp1y, cp2x, cp2y, x2, y2, t) {
  const mt = 1 - t, mt2 = mt * mt, mt3 = mt2 * mt;
  const t2 = t * t, t3 = t2 * t;
  return {
    x: mt3 * x1 + 3 * mt2 * t * cp1x + 3 * mt * t2 * cp2x + t3 * x2,
    y: mt3 * y1 + 3 * mt2 * t * cp1y + 3 * mt * t2 * cp2y + t3 * y2,
  };
}

function quadBezierAt(x1, y1, cpx, cpy, x2, y2, t) {
  const mt = 1 - t;
  return {
    x: mt * mt * x1 + 2 * mt * t * cpx + t * t * x2,
    y: mt * mt * y1 + 2 * mt * t * cpy + t * t * y2,
  };
}

/**
 * Approximate an SVG arc segment as a polyline.
 * Uses endpoint-to-center parameterization.
 */
function approximateArc(seg, n) {
  const { x1, y1, rx: _rx, ry: _ry, rotation, largeArc, sweep, x2, y2 } = seg;
  let rx = Math.abs(_rx), ry = Math.abs(_ry);
  if (rx === 0 || ry === 0) {
    return [{ x: x1, y: y1 }, { x: x2, y: y2 }];
  }

  const phi = (rotation * Math.PI) / 180;
  const cosPhi = Math.cos(phi), sinPhi = Math.sin(phi);

  const dx2 = (x1 - x2) / 2, dy2 = (y1 - y2) / 2;
  const x1p = cosPhi * dx2 + sinPhi * dy2;
  const y1p = -sinPhi * dx2 + cosPhi * dy2;

  let lambda = (x1p * x1p) / (rx * rx) + (y1p * y1p) / (ry * ry);
  if (lambda > 1) { rx *= Math.sqrt(lambda); ry *= Math.sqrt(lambda); }

  const num = Math.max(0, rx * rx * ry * ry - rx * rx * y1p * y1p - ry * ry * x1p * x1p);
  const den = rx * rx * y1p * y1p + ry * ry * x1p * x1p;
  let sq = den > 0 ? Math.sqrt(num / den) : 0;
  if (largeArc === sweep) sq = -sq;

  const cxp = sq * (rx * y1p) / ry;
  const cyp = sq * -(ry * x1p) / rx;

  const cx = cosPhi * cxp - sinPhi * cyp + (x1 + x2) / 2;
  const cy = sinPhi * cxp + cosPhi * cyp + (y1 + y2) / 2;

  const theta1 = Math.atan2((y1p - cyp) / ry, (x1p - cxp) / rx);
  let dtheta = Math.atan2((-y1p - cyp) / ry, (-x1p - cxp) / rx) - theta1;

  if (sweep && dtheta < 0) dtheta += 2 * Math.PI;
  if (!sweep && dtheta > 0) dtheta -= 2 * Math.PI;

  const pts = [];
  for (let i = 0; i <= n; i++) {
    const t = theta1 + (i / n) * dtheta;
    const xr = rx * Math.cos(t), yr = ry * Math.sin(t);
    pts.push({
      x: cosPhi * xr - sinPhi * yr + cx,
      y: sinPhi * xr + cosPhi * yr + cy,
    });
  }
  return pts;
}
